doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
```

Comment:
I think the options are quite different --- one allows opt-out with array_ufunc, and the other not.

As an aside, the ArrayLike example in the gist uses an unconditional nested ufunc dispatch --- which puts it in the type hierarchy above all other types, which probably is not intended. I think this can be avoided by first making type checks (e.g. accept ndarray, ArrayLike and otherwise return NotImplemented) --- or, delegate also the type checking by calling `self.value.__array_ufunc__(...)`.  I don't really see this as a question about a public vs private API, as the special method is not really private API, and you should be able to delegate it the same way as `__mul__` et al.